// sb-api/lib/lightning/clightning.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var assert_1 = __importDefault(require("assert"));
var debug_1 = __importDefault(require("debug"));
var fs_1 = __importDefault(require("fs"));
var lightning_client_1 = __importDefault(require("lightning-client"));
var os_1 = __importDefault(require("os"));
var path_1 = __importDefault(require("path"));
var util_1 = __importDefault(require("util"));
var uuid_1 = __importDefault(require("uuid"));
var debug = debug_1.default('sb-api:lightning:clightning');
var lstat = util_1.default.promisify(fs_1.default.lstat);
var CLightningImpl = /** @class */ (function () {
    function CLightningImpl(rpcPath) {
        var _this = this;
        this.getPreimage = function (invoice) {
            debug("Getting preimage for invoice " + invoice);
            var preimage = _this.preimages[invoice];
            if (!preimage) {
                return Promise.reject(new Error("Could not find preimage for invoice " + invoice));
            }
            else {
                return Promise.resolve(preimage);
            }
        };
        this.receive = function (description) {
            if (description === void 0) { description = ''; }
            var uniqueLabel = uuid_1.default(); // c-lightning requires all invoice labels to be unique
            // c-lightning expects the string 'any' to make an invoice without amount specified
            var amountMsat = 'any';
            debug("Getting a new invoice with " + (description ? "description: " + description : 'no description'));
            return _this.client.invoice(amountMsat, uniqueLabel, description).then(function (_a) {
                var bolt11 = _a.bolt11;
                return bolt11;
            });
        };
        this.send = function (invoice) {
            debug("Paying invoice " + invoice);
            return _this.client
                .pay(invoice)
                .then(function (res) {
                debug("Result of paying invoice " + invoice.slice(0, 10) + ": %O", res);
                _this.preimages[invoice] = res.payment_preimage;
                return res;
            })
                .catch(function (err) {
                debug("Error: when paying invoice: %O", err);
                throw err;
            });
        };
        /** Maps over paid invoices with their corresponding preimage */
        this.preimages = {};
        this.getInfo = function () { return _this.client.getinfo(); };
        debug("Connecting to c-lightning with rpcPath " + rpcPath);
        this.client = new lightning_client_1.default(rpcPath);
    }
    return CLightningImpl;
}());
var DEFAULT_DATADIR = path_1.default.join(os_1.default.homedir(), '.lightning');
exports.CLightning = function (rpcPath) {
    if (rpcPath === void 0) { rpcPath = DEFAULT_DATADIR; }
    return __awaiter(_this, void 0, void 0, function () {
        var lstats, isDirectory, client, info;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    debug("rpcPath: " + rpcPath);
                    return [4 /*yield*/, lstat(rpcPath)];
                case 1:
                    lstats = _a.sent();
                    isDirectory = lstats.isDirectory;
                    assert_1.default(isDirectory, rpcPath + " is not a directory!");
                    client = new CLightningImpl(rpcPath);
                    debug('Trying to connect to client');
                    return [4 /*yield*/, client.getInfo()];
                case 2:
                    info = _a.sent();
                    debug('Succeeded!');
                    debug('getinfo output: %o', info);
                    return [2 /*return*/, client];
            }
        });
    });
};
