// sb-api/lib/types/exchange/spot/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var t = __importStar(require("io-ts"));
var books_1 = require("./books");
var tickers_1 = require("./tickers");
var trades_1 = require("./trades");
var ExchangeSpotTypes = /** @class */ (function () {
    function ExchangeSpotTypes() {
    }
    ExchangeSpotTypes.DataTypes = {
        trades: trades_1.ExchangeSpotTradesTypes,
        books: books_1.ExchangeSpotBooksTypes,
        tickers: tickers_1.ExchangeSpotTickersTypes,
    };
    ExchangeSpotTypes.Exchange = t.keyof({
        coinbase: t.null,
        binance: t.null,
        bitfinex: t.null,
        gemini: t.null,
        bitstamp: t.null,
        kraken: t.null,
    });
    return ExchangeSpotTypes;
}());
exports.ExchangeSpotTypes = ExchangeSpotTypes;
