// sb-api/lib/types/exchange/futures/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var t = __importStar(require("io-ts"));
var books_1 = require("./books");
var tickers_1 = require("./tickers");
var trades_1 = require("./trades");
var ExchangeFuturesTypes = /** @class */ (function () {
    function ExchangeFuturesTypes() {
    }
    ExchangeFuturesTypes.DataTypes = {
        trades: trades_1.ExchangeFuturesTradesTypes,
        books: books_1.ExchangeFuturesBooksTypes,
        tickers: tickers_1.ExchangeFuturesTickersTypes,
    };
    ExchangeFuturesTypes.Exchange = t.keyof({
        kraken: t.null,
    });
    ExchangeFuturesTypes.Interval = t.keyof({
        monthly: null,
        quarterly: null,
        perpetual: null,
    });
    ExchangeFuturesTypes.ALL_EXCHANGE_FUTURES_TYPES = [];
    return ExchangeFuturesTypes;
}());
exports.ExchangeFuturesTypes = ExchangeFuturesTypes;
